### ISSUE TITLE
CrabSpectrum as factory function

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1788]
I was using `gammapy.spectrum.CrabSpectrum` again today, and was puzzled by the API. It's a class with only one method `__init__` which is a factory function, that stores the return value on the instance since `__init__` can't return things:
http://docs.gammapy.org/dev/_modules/gammapy/spectrum/crab.html#CrabSpectrum
So callers always do `CrabSpectrum('hess').model`, never use the `CrabSpectrum` object for anything else.

Looking over #777 it seems there might be a use case to have the class and extend it, or there might not.

This is a reminder issue for v0.9 to discuss this point again and to consider changing this to a factory function `crab_spectrum` that does what `CrabSpectrum.__init__` does now.

We could also consider moving this to the `gammapy.spectrum.models` namespace.

**Comment by adonath** [comment]
I think introducing a factory method is just fine. The weak motivation for having the `CrabSpectrum` class was to keep a reference for the model in the `.reference` attribute. But it's probably sufficient to keep the reference e.g. by the variable name...

**Comment by cdeil** [comment]
Since Python is dynamic we could just do this?
```
model = ...
model.meta = {"reference": reference}
return model
```
I think `reference` is sometimes a (bad) model parameter name for the reference energy, and we now expose it on the model class. So we have to be careful concerning namespace collisions?

I'll wait for @registerrier to comment, then send a PR for this soon.

**Comment by adonath** [comment]
@deil My point was even that we probably don't need to define this meta data for now. Users can take care of it themselves, by choosing good variable names: `hess_crab_pl = crab_spectrum("hess_pl")`

**Comment by cdeil** [comment]
OK, so the task here is to refactor `CrabSpectrum` into a function,  and to not attache the reference to the returned model as is currently done [here](https://github.com/gammapy/gammapy/blob/50b2b8ae43456011af0575d45dba0c5a9aeae1c8/gammapy/spectrum/crab.py#L137).

Remaining decision needed IMO is whether the function name should be `crab_spectrum` or `make_crab_spectrum`, and whether to expose it in the `gammapy.spectrum` namespace or in the `gammapy.spectrum.models` namespace.

My recommendation would be to call it `make_crab_spectrum`, to expose it in `gammapy.spetrum.models`, and to keep the code location and filename as-is: `gammapy/spectrum/crab.py`.

(similarly for #2213 I would suggest to call it `make_cosmic_ray_spectrum`)

@adonath @registerrier - What's your preference?

If anyone is interested to work on this task, please leave a comment or get in touch on Slack.

**Comment by adonath** [comment]
I agree it should be exposed in the `gammapy.spectrum.models` namespace. At some point we might think about adding one more sub-package i.e. `spectrum/models/crab.py` and `spectrum/models/core.py` or similar, just to avoid the `spectrum/models.py` file grows to large (it's already ~1600 lines...)

I don't have a strong opinion on the naming scheme, but my first proposal would be `create_crab_spectral_model` and `create_cosmic_ray_spectral_model`, because for the factory methods in `gammapy.maps` we already use "create" and I think it also good to mention "model" in the name of the methods.

**Comment by cdeil** [comment]
For the filename / namespace: now I see it's actually problematic to keep the function in `crab.py` to expose it in `models.py` because putting `from .crab import make_crab_spectrum` in `models.py` and listing it there in `__all__` would mean delayed imports to avoid circular imports in `crab.py`. Aargh.

Let's not move the code for now, it should stay in `gammapy/models.crab.py` and be exposed in `gammapy.spectrum` as-is, and then in the future as a separate task we probably change to a `gammapy/spectrum/models/` package with multiple files there, like `crab.py` or `core.py`.

We have plenty of functions with [create](https://docs.gammapy.org/0.12/search.html?q=create&check_keywords=yes&area=default) and plenty with [make](https://docs.gammapy.org/0.12/search.html?q=make&check_keywords=yes&area=default#) currently. Let's not discuss this now and here which way to unify, but say that whoever implements this task gets to choose for now and we harmonise throughout Gammapy later?

**Comment by adonath** [comment]
Resolved in #2278.
